geode-core/src/main/java/org/apache/geode/management/internal/api/SmartClusterManagementServiceProviderFactory.java

```diff
@@ -0,0 +1,128 @@
+ * agreements. See the NOTICE file distributed with this work for additional information regarding
+ * copyright ownership. The ASF licenses this file to You under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance with the License. You may obtain a
+ * copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software distributed under the License
+ * is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
+ * or implied. See the License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+package org.apache.geode.management.internal.api;
+
+import java.util.Arrays;
+import java.util.List;
+import java.util.Set;
+
+import org.apache.commons.lang3.StringUtils;
+import org.apache.logging.log4j.Logger;
+
+import org.apache.geode.annotations.Immutable;
+import org.apache.geode.cache.Cache;
+import org.apache.geode.cache.CacheFactory;
+import org.apache.geode.cache.client.ClientCache;
+import org.apache.geode.cache.client.ClientCacheFactory;
+import org.apache.geode.cache.execute.FunctionException;
+import org.apache.geode.cache.execute.FunctionService;
+import org.apache.geode.cache.execute.ResultCollector;
+import org.apache.geode.distributed.internal.InternalLocator;
+import org.apache.geode.distributed.internal.membership.InternalDistributedMember;
+import org.apache.geode.internal.cache.GemFireCacheImpl;
+import org.apache.geode.internal.logging.LogService;
+import org.apache.geode.management.api.ClusterManagementService;
+import org.apache.geode.management.client.ClusterManagementServiceProvider;
+import org.apache.geode.management.internal.ClusterManagementClient;
+import org.apache.geode.management.internal.cli.domain.MemberInformation;
+import org.apache.geode.management.internal.cli.functions.GetMemberInformationFunction;
+import org.apache.geode.management.spi.BasicClusterManagementProviderFactory;
+import org.apache.geode.management.spi.ClusterManagementServiceProviderFactory;
+
+/**
+ * An implementation of {@link ClusterManagementServiceProviderFactory} which can be used in any
+ * context where Geode is running (client, server or locator). It will attempt to determine the
+ * address of the {@code ClusterManagementService} when using the {@code create()} call. Otherwise
+ * an explicit can also be used.
+ */
+public class SmartClusterManagementServiceProviderFactory extends
+    BasicClusterManagementProviderFactory {
+
+  @Immutable
+  private static final GetMemberInformationFunction MEMBER_INFORMATION_FUNCTION =
+      new GetMemberInformationFunction();
+
+  private static final Logger logger = LogService.getLogger();
+
+  @Override
+  public List<String> supportedContexts() {
+    return Arrays.asList(ClusterManagementServiceProvider.GEODE_CLIENT_CONTEXT,
```

Comment:
should this also have LOCATOR_CONTEXT? or if the create() method supports both "LOCATOR" and "SERVER" mode, then we only need one string for this.